src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/simple/SimpleJoinOrderingQueryOptimizer.java

```diff
@@ -41,14 +48,26 @@ public PhysicalPlan optimize( final LogicalPlan initialPlan ) throws QueryOptimi
     }
 
     public PhysicalPlan optimizePlan( final PhysicalPlan plan ) throws QueryOptimizationException {
+        final PhysicalPlan[] optSubPlans = getOptimizedSubPlans(plan);
+
         if ( hasMultiwayJoinAsRoot(plan) ){
-            final PhysicalPlan[] subplans = getOptimizedSubPlans(plan);
-            return joinPlanOptimizer.determineJoinPlan(subplans);
+            return joinPlanOptimizer.determineJoinPlan(optSubPlans);
         }
-        else {
-        	// TODO incorrect!! There may be multiway joins inside the subplans of the given plan.
+        else if (hasNullaryOpAsRoot(plan)){
             return plan;
         }
+        else if ( hasUnaryOpAsRoot(plan) ){
```

Comment:
Instead of calling these helper methods (`hasNullaryOpAsRoot`, etc), I think it is safe to distinguish the cases based on the number of subplans (i.e., `plan.numberOfSubPlans()`). That way, the code becomes a bit more efficient because it avoids doing the casts to `PhysicalOperatorForLogicalOperator` that are done in each of these helper methods.